design/cassandra/cluster-creation.md

```diff
@@ -0,0 +1,242 @@
+            memory: "4Gi"
+          limits:
+            cpu: "2000m"
+            memory: "4Gi"
+        - name: sidecar
+          requests:
+            cpu: "200m"
+            memory: "200M"
+          limits:
+            cpu: "200m"
+            memory: "200M"
+```
+
+2. **Controller** is informed of a new Cluster CRD instance. For each (dc,rack) it creates:
+    1. A StatefulSet (see Appendix A) 
+    2. A Headless Service (needed for the StatefulSet)
+    3. ClusterIP Services that serve as static IPs for Pods. Labeled accordingly if they are seeds. Each Service is named after the Pod that uses it.
+
+3. **StatefulSet Pod** starts and it starts the Cassandra and Sidecar containers. Then the following is done to customize config options for the instance:
+    1. **Cassandra**: copy default config files from `/etc/cassandra` to `/mnt/shared/default`.
+    2. **Sidecar**: Check for already created `ClusterIP` service by the Pod Name. Note if you are a Seed in the ClusterIP Service. Poll for the `default` folder, then merge default configuration with the operator's options (see Appendix B) and the user provided options. Write the result under `/mnt/shared/custom`
+    3. **Cassandra**: poll for the `custom` folder, then copy custom config files to `/etc/cassandra` and start the instance.
+
+## Concerns
+
+* This choreography of actions seems a bit excessive.
+* **Reasoning:** it is necessary in order to achieve 2 things:
+    1. Keep the Cassandra instance and the Sidecar isolated in different containers. We don't want our instance crashing because of the sidecar. 
+    2. Ability to separately update sidecar and cassandra, because they live in different containers and use different images.
+
+## Appendix A
+
+The StatefulSet that will be created by the controller is the following:
+
+``` yaml
+apiVersion: apps/v1
+kind: StatefulSet
+metadata: 
+  name: <sts-name>
+  namespace: <sts-name>
+  labels:
+    # Kubernetes recommended labels
+    app.kubernetes.io/name: cassandra
+    app.kubernetes.io/managed-by: rook 
+    # rook operator labels
+    cassandra.rook.io/cluster: <cluster>
+    cassandra.rook.io/datacenter: <datacenter>
+    cassandra.rook.io/rack: <rack>
+spec:
+  replicas: <instances>
+  serviceName: <sts-name>-hs
+  selector:
+    matchLabels:
+      cassandra.rook.io/cluster: <cluster>
+      cassandra.rook.io/datacenter: <datacenter>
+      cassandra.rook.io/rack: <rack>
+  template:
+    metadata:
+      labels:
+        cassandra.rook.io/cluster: <cluster>
+        cassandra.rook.io/datacenter: <datacenter>
+        cassandra.rook.io/rack: <rack>
+    spec: 
+      volumes: 
+        - name: shared
+          emptyDir: {}
+      containers:
+        - name: cassandra
+          image: "<repository>:<version>"
+          imagePullPolicy: IfNotPresent
+          command:
+            - "/bin/bash"
+            - "-c"
+            - |
+              set -e
+              CASSANDRA_CONFIG="/etc/cassandra"
```

Comment:
i mentioned this in #2207, but can this be done in the containers entrypoint instead?